### ISSUE TITLE
Errorprone separate testing project

### ISSUE THREAD

**Issue by sergiofigueras** [google/error-prone#431]
Hello guys,

In our current work, we should develop several different BugCheckers for Errorprone. So, I've created a separate project from error_prone_core containing some of Errorprone test core classes.

I think that it's better because other developers can create tests out of error_prone_core. In our case, for example, we should use errorprone with some of our BugCheckers as a plugin, and it's better if we distribute only a simple "errorprone-testing" project instead of tell to developers "hey, clone this repos and create bugcheckers inside of it".

What do you think guys? If you want, I can start to work on this errorprone separate test project.

**Comment by cushon** [comment]
Enforcing a cleaner separation between the built-in checks and everything else in error_prone_core sounds good too me. Would creating a separate maven module be sufficient, or did you have something else in mind?

**Comment by sergiofigueras** [comment]
I think that for now, we could just create a separate maven module to do this. That would be something like "error_prone_test_suite" that would have classes like `CompilationTestHelper` that would help developers to use. Am I understanding your idea correctly? What do you think?

**Comment by cushon** [comment]
Yep, that's what I was thinking. `CompilationTestHelper` depends on core classes like `BaseErrorProneCompiler`, though, so we'd either need three pieces (core, test, and built in checks), or have one piece that combines core and test, and split out the built-in checks.

**Comment by sergiofigueras** [comment]
Ok @cushon , I can handle it. How do you guys want to come up with this? What's the development process that you think that is better for this?

**Comment by cushon** [comment]
> Also, I would think about separate the compiler itself in another maven module too.

Can you elaborate on this part? I was thinking that 'built-in checks' would include only the checks under `/bugpatterns/`, and everything else that wasn't test-only would stay in 'core'. The compiler has a lot of dependencies in core, so it's not clear to me what we'd gain from putting it in a separate module.

> I can handle it. How do you guys want to come up with this? What's the development process that you think that is better for this?

Thanks, let me confirm that everyone likes the new layout first. It's also going to require changes to the internal build, so it might be easier if I handled that and then synced it to open source.

**Comment by cushon** [comment]
There's some progress on this in 4dfc65ad93fe42f0fda2577235a01ab68a4901bf.
